### PR TITLE
Instrument `Spec<T>` lambdas for compatibility with the configuration cache

### PR DESCRIPTION
**Pull request by bamboo**

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java [hunk 1]
```diff
@@ -272,6 +310,15 @@ public void visitInvokeDynamicInsn(String name, String descriptor, Handle bootst
...
             }
         }
+
+        private boolean isGradleLambdaDescriptor(String descriptor) {
+            return isLambdaDescriptorOf(Action.class, descriptor)
+                || isLambdaDescriptorOf(Spec.class, descriptor);
+        }
+
+        private boolean isLambdaDescriptorOf(Class<?> clazz, String descriptor) {
+            return descriptor.endsWith(")" + getType(clazz).getDescriptor());
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java:320]: maybe turn this into some constants, as this is called many times.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java [hunk 2]
```diff
@@ -133,136 +148,159 @@ public void visitEnd() {
...
+
         @Override
         public void visitInvokeDynamicInsn(String name, String descriptor, Handle bootstrapMethodHandle, Object... bootstrapMethodArguments) {
-            if (descriptor.endsWith(")" + Type.getType(Action.class).getDescriptor()) && bootstrapMethodHandle.getOwner().equals(Type.getType(LambdaMetafactory.class).getInternalName()) && bootstrapMethodHandle.getName().equals("metafactory")) {
-                Handle altMethod = new Handle(Opcodes.H_INVOKESTATIC,
-                    Type.getType(LambdaMetafactory.class).getInternalName(),
+            if (isGradleLambdaDescriptor(descriptor) && bootstrapMethodHandle.getOwner().equals(getType(LambdaMetafactory.class).getInternalName()) && bootstrapMethodHandle.getName().equals("metafactory")) {
+                Handle altMethod = new Handle(
+                    H_INVOKESTATIC,
+                    getType(LambdaMetafactory.class).getInternalName(),
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java:299]: we should probably turn this and the method descriptor below into constants, similar to what happens for other descriptors and type names (I should have done that initially but forgot).

**Comment by adammurdoch** [comment]
Also, when changing the instrumentation, we need to increment the instrumentation version in `applyConfigurationTo()`, so that we don't reuse things that were previously instrumented without support for `Spec`.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java (hunk 1 above)
Comment (bamboo) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java]: Done.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java (hunk 2 above)
Comment (bamboo) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java]: Done.

**Comment by bamboo** [comment]
@bot-gradle test this

**Comment by bamboo** [comment]
Thanks for the review, @adammurdoch. All your suggestions have been applied. PTAL.

**Pull request merged**